File: app/src/main/java/com/wkz/viewpager/transformer/AbstractBaseTransformer.java
package com.wkz.viewpager.transformer;

import androidx.annotation.NonNull;
import androidx.viewpager.widget.ViewPager;

import android.view.View;

import com.nineoldandroids.view.ViewHelper;

/**
 * @desc 基础转换效果
 * @date 2017-06-30
 */
public abstract class AbstractBaseTransformer implements ViewPager.PageTransformer {

    @Override
    public void transformPage(@NonNull View page, float position) {
        onPreTransform(page, position);
        onTransform(page, position);
        onPostTransform(page, position);
    }

    protected void onPreTransform(View page, float position) {
        float pageWidth = page.getWidth();
        ViewHelper.setRotationX(page, 0F);
        ViewHelper.setRotationY(page, 0F);
        ViewHelper.setRotation(page, 0F);
        ViewHelper.setScaleX(page, 1F);
        ViewHelper.setScaleY(page, 1F);
        ViewHelper.setPivotX(page, 0F);
        ViewHelper.setPivotY(page, 0F);
        ViewHelper.setTranslationX(page, isPagingEnabled() ? 0F : -pageWidth * position);
        ViewHelper.setTranslationY(page, 0F);

        if (hideOffscreenPages()) {
            ViewHelper.setAlpha(page, position <= -1 || position >= 1 ? 0F : 1F);
        } else {
            ViewHelper.setAlpha(page, 1F);
        }
    }

    protected void onPostTransform(View page, float position) {
    }

    protected boolean hideOffscreenPages() {
        return true;
    }

    protected boolean isPagingEnabled() {
        return false;
    }

    protected float max(float val, float max) {
        return val < max ? max : val;
    }

    protected abstract void onTransform(View page, float position);
}

File: view-pager-transformer/src/main/java/com/fphoenixcorneae/view_pager_transformer/AbstractBaseTransformer.kt
package com.fphoenixcorneae.view_pager_transformer

import android.view.View
import androidx.viewpager.widget.ViewPager.PageTransformer
import com.nineoldandroids.view.ViewHelper

/**
 * @desc 基础转换效果
 * @date 2017-06-30
 */
abstract class AbstractBaseTransformer : PageTransformer {
    override fun transformPage(page: View, position: Float) {
        onPreTransform(page, position)
        onTransform(page, position)
        onPostTransform(page, position)
    }

    protected fun onPreTransform(page: View, position: Float) {
        val pageWidth = page.width.toFloat()
        ViewHelper.setRotationX(page, 0f)
        ViewHelper.setRotationY(page, 0f)
        ViewHelper.setRotation(page, 0f)
        ViewHelper.setScaleX(page, 1f)
        ViewHelper.setScaleY(page, 1f)
        ViewHelper.setPivotX(page, 0f)
        ViewHelper.setPivotY(page, 0f)
        ViewHelper.setTranslationX(page, if (isPagingEnabled) 0f else -pageWidth * position)
        ViewHelper.setTranslationY(page, 0f)
        if (hideOffscreenPages()) {
            ViewHelper.setAlpha(page, if (position <= -1 || position >= 1) 0f else 1f)
        } else {
            ViewHelper.setAlpha(page, 1f)
        }
    }

    protected fun onPostTransform(page: View, position: Float) {}

    protected fun hideOffscreenPages(): Boolean {
        return true
    }

    protected open val isPagingEnabled: Boolean = false

    protected fun max(`val`: Float, max: Float): Float {
        return kotlin.math.max(`val`, max)
    }

    protected abstract fun onTransform(page: View, position: Float)
}